controllers/storageconsumer/storageconsumer_controller.go

```diff
@@ -52,11 +84,439 @@ func (r *StorageConsumerReconciler) Reconcile(ctx context.Context, request recon
 
 	r.Log.Info("Reconciling StorageConsumer.", "StorageConsumer", klog.KRef(request.Namespace, request.Name))
 
-	// TODO: Implement complete functionality of the controller
+	// Initalize the reconciler properties from the request
+	r.initReconciler(request)
+
+	if err := r.get(r.storageConsumer); err != nil {
+		if errors.IsNotFound(err) {
+			r.Log.Info("No StorageConsumer resource.", "StorageConsumer", klog.KRef(r.storageConsumer.Namespace, r.storageConsumer.Name))
+			// Request object not found, could have been deleted after reconcile request.
+			// Owned objects are automatically garbage collected. For additional cleanup logic use finalizers.
+			// Return and don't requeue
+			return reconcile.Result{}, nil
+		}
+		// Error reading the object - requeue the request.
+		r.Log.Error(err, "Failed to retrieve StorageConsumer.", "StorageConsumer", klog.KRef(r.storageConsumer.Namespace, r.storageConsumer.Name))
+		return reconcile.Result{}, err
+	}
+
+	// Reconcile changes to the cluster
+	result, reconcileError := r.reconcilePhases()
+
+	// Apply status changes to the StorageConsumer
+	statusError := r.Client.Status().Update(r.ctx, r.storageConsumer)
+	if statusError != nil {
+		r.Log.Info("Could not update StorageConsumer status.", "StorageConsumer", klog.KRef(r.storageConsumer.Namespace, r.storageConsumer.Name))
+	}
+
+	// Reconcile errors have higher priority than status update errors
+	if reconcileError != nil {
```

Comment:
Because we want to guarantee that the status is written to k8s even when reconcile have errors.
So we cannot handle the error before writing the status